### DISCUSSION TITLE
Can not read ebusd over TCP from python

### DISCUSSION THREAD

**Discussion by renco33** [john30/ebusd#1345] [Q&A]
somewhat unrelated question but I'm really stuck...
I can read ebucd from shell with netcat command:
 ```
user@raspberrypi:~ $ echo r -c 700 AdaptHeatCurve | nc -q 1 127.0.0.1  8888
no
```
The response  received and  "no" is correct.
But trying to do same thing from python - just returns empty string. And seems no connection errors. Here is code:
```
#function to read ebusd value
def readValue(circuit, field):
    content = "read -c 700 AdaptHeatCurve"
    print("debug1")
    sock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
    try: 
        sock.connect(('127.0.0.1', 8888))
    except socket.error as msge:
        print ("connection failed error :", msge)
    print("debug2" )  
    try:   
        sock.sendall(content.encode())
    except socket.error as msge:
        print ("sendall failed error :", msge)       
    print("debug3" )
    time.sleep(5)
    print("debug4" )
    sock.shutdown(socket.SHUT_WR)
    print("debug5" )   
    res = ""
        
    while True:
        data = sock.recv(1024)
        if (not data):
            break
         
    print("readvalue: ",repr(data))
    sock.close()
```

The output of above:
```
(myTestServer) user@raspberrypi:~ $ python3 myTestServer/testServer.py
debug1
debug2
debug3
debug4
debug5
readvalue:  b''
```

Anybopdy could point what is wrong? ( I'm novice - dumb things possible)

**Comment by limad** [comment]
Hi
defaut port to connect socket is 8888
But be sure you running ebusd with the appropriate config ''--port=8888"

**Comment by ningyichensha** [comment]
I run ebusd for a ochsner heatpump.
Maybe my solution using python works for you as well.

I use hex statements to pass the commands to the ebus
For example in the raspberry terminal: ebusctl hex 1006230603b200480100
That returns "done" and changes the mode for the hk1 to automatic (third last byte = 1)
To make ebusctl accept hex input, add --enablehex in file: /etc/default/ebusd.
In my case the line is: EBUSD_OPTS="--enablehex -d ens:/dev/ttyAMA0 --latency=50"

To find out which hex code need to be passed to the ebus, I looked at the log file at
/media/user/rootfs/var/log/ebusd.log

When you submit your statement, that returns "no" a couple of times, then you should find the hex code in the log.
From the hex code you copy everything without the source address, which is in the first and second byte.
Now you try the statement: ebusctl hex "your hexcode"
That should return the desired answer "no".

Next step is the implementation in python:
My code to change the mode of the heatpump is this:
import subprocess
...
befehl = subprocess.run("ebusctl hex 1006230603b200480100", shell = True ,capture_output=True)
antwort = befehl.stdout
out = antwort.decode()
out = out.strip()
if out == "done":
.....

**Comment by jvanoosterhout** [comment]
Interesting, I tend to do the same thing when my ebus runs. So I cannot test your code but it seems you close the socket after debug 4 but try to read from it only after debug 5. Hopefully changing the order works out for you.

**Comment by jvanoosterhout** [comment]
I found the awnser to the issue: you need to include a new line (\n) in the data send in order to conclude the command send. Here is a minimal example to request “info”, it includs some other fixes like accounting for message longer than 1024 bytes. 
```
import socket
import time

def recvall(sock):
    BUFF_SIZE = 512 
    data = b''
    while True:
        part = sock.recv(BUFF_SIZE)
        data += part
        print('received a part')
        if len(part) < BUFF_SIZE:
            # either 0 or end of data
            break
    return data

HOST = "localhost"  # The server's hostname or IP address
PORT = 8888  # The port used by the server

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.settimeout(5)
    s.connect((HOST, PORT))
    s.send(b"info\n")
    data = recvall(s)
 
print("readvalue: ",repr(data))
s.close()
```
